src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_NonNullTypes.cs

```diff
@@ -0,0 +1,297 @@
+                Diagnostic(ErrorCode.ERR_DuplicateNameInNS).WithArguments("NonNullTypesAttribute", "System.Runtime.CompilerServices").WithLocation(1, 1)
+                );
+
+            system = (INamespaceSymbol)comp.GlobalNamespace.GetMember("System");
+            Equal(NamespaceKind.Compilation, system.NamespaceKind);
+        }
+
+        [Fact]
+        public void CannotReferenceNonNullTypesAttributesFromADifferentAssembly_Public()
+        {
+            var reference = CreateCompilation(@"
+namespace System.Runtime.CompilerServices
+{
+    internal class NonNullTypesAttribute : System.Attribute { public NonNullTypesAttribute(bool flag = true) { } }
+}
+
+[System.Runtime.CompilerServices.NonNullTypes]
+public class TestType1 { }
+
+[System.Runtime.CompilerServices.NonNullTypesAttribute(false)]
+public class TestType2 { }
+");
+            False(reference.GetMember("TestType1").GetAttributes().Single().AttributeClass.IsImplicitlyDeclared);
+            False(reference.GetMember("TestType2").GetAttributes().Single().AttributeClass.IsImplicitlyDeclared);
+
+            var code = "[module: System.Runtime.CompilerServices.NonNullTypes]";
+
+            var comp = CreateCompilation(code, references: new[] { reference.ToMetadataReference() });
+            comp.VerifyDiagnostics();
+
+            True(comp.SourceModule.GetAttributes().Single().AttributeClass.IsImplicitlyDeclared);
+        }
+
+        [Fact]
+        public void NonNullTypesAttributeInSourceCanBeUsed()
+        {
+            var code = @"
+// This should trigger generating another NonNullTypesAttribute
+[module: System.Runtime.CompilerServices.NonNullTypes]
+
+namespace System.Runtime.CompilerServices
+{
+    public class NonNullTypesAttribute : System.Attribute { public NonNullTypesAttribute(bool flag = true) { } }
+}";
+
+            var comp = CreateCompilation(code);
+            comp.VerifyEmitDiagnostics();
+            True(comp.SourceModule.NonNullTypes);
+            False(comp.SourceModule.GetAttributes().Single().AttributeClass.IsImplicitlyDeclared);
+        }
+
+        [Fact]
+        public void NonNullTypesAttributeInReferencedModuleCausesAmbiguity()
+        {
+            var module = CreateCompilation(options: TestOptions.ReleaseModule, assemblyName: "testModule", source: @"
+namespace System.Runtime.CompilerServices
+{
+    public class NonNullTypesAttribute : System.Attribute { public NonNullTypesAttribute(bool flag = true) { } }
+}");
+
+            var moduleRef = ModuleMetadata.CreateFromImage(module.EmitToArray()).GetReference();
+
+            var code = @"
+class Test
+{
+    public void M(in int p)
+    {
+        // This should trigger generating another NonNullTypesAttribute
+    }
+}";
+
+            CreateCompilation(code, references: new[] { moduleRef }).VerifyEmitDiagnostics(
+                // error CS0101: The namespace 'System.Runtime.CompilerServices' already contains a definition for 'NonNullTypesAttribute'
```

Comment:
How is this case different from `CompilerShouldIgnorePublicNonNullTypesAttributesInReferencedAssemblies` where no error is reported? Consider at least adding a comment explaining the behavior since this looks like a significant issue.

---
In reply to: [214478275](https://github.com/dotnet/roslyn/pull/29180#discussion_r214478275) [](ancestors = 214478275,214437799)